agent/cache-types/streaming_health_services.go

```diff
@@ -240,3 +288,34 @@ func (s *healthView) Result(index uint64) (interface{}, error) {
 func (s *healthView) Reset() {
 	s.state = make(map[string]structs.CheckServiceNode)
 }
+
+// serviceTagEvaluator implements the filterEvaluator to perform filtering
+// by service tags. bexpr can not be used at this time, because the filtering
+// must be case insensitive for backwards compatibility. In the future this
+// may be replaced with bexpr once case insensitive support is added.
+type serviceTagEvaluator struct {
+	tags []string
+}
+
+func (m serviceTagEvaluator) Evaluate(data interface{}) (bool, error) {
+	csn, ok := data.(structs.CheckServiceNode)
+	if !ok {
+		return false, fmt.Errorf("unexpected type %T for structs.CheckServiceNode filter", data)
+	}
+	for _, tag := range m.tags {
+		if !serviceHasTag(csn.Service, tag) {
+			// If any one of the expected tags was not found, filter the service
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func serviceHasTag(sn *structs.NodeService, tag string) bool {
+	for _, t := range sn.Tags {
```

Comment:
I'm pretty sure the answer is no, but do we ever need to handle the backcompat case where `sn.Tag` is set? I think not since this would have come from Consul's state store and every CSN from a version recent enough to support streaming would only use the plural field right?

Reply:
That case is handled in this PR, here: https://github.com/hashicorp/consul/pull/9703/files/d5cc206e3963ddd910ee34b8612658d85f370d38#diff-97d0de9c08776c4ff05f75254d7319f757e5b56f78b428d0b2574ee43b77bb9dR210